BiometricLocker/BiometricLocker.swift

```diff
@@ -9,6 +9,189 @@
+                // If app goes into background, we start the clocks to lock the app.
+                self.lock()
+            }
+
+            NotificationCenter.default.addObserver(forName: .UIApplicationWillTerminate, object: nil, queue: .main) { _ in
+                // If the app is killed, lock it instantly.
+                self.lock(.now)
+            }
+        }
+    }
+
+    /**
+     Tells the biometric locker to lock the app. Defaults to the currently defined time allowance (5min by default).
+
+     - Parameter when: An enum defining when the app should lock. Now, after the time allowance, or after a custom time interval.
+     */
+    public func lock(_ when: LockingTime = .afterTimeAllowance) {
+        // if we are already deactivated, return. Otherwise you can just kill the app and try again to bypass touch ID.
+        if self.isLocked { return }
+
+        let date: Date
+        switch when {
+        case .now:
+            date = Date.distantPast
+        case .afterTimeAllowance:
+            date = Date()
+        case .afterTimeInterval(let interval):
+            // We subtract our `unlockedTimeAllowance` here, to offset when checking again
+            // inside `isLocked`, otherwise it would be up to the user to know how we implement this internally
+            // and manually calculate the time difference. 
+            date = Date(timeIntervalSinceNow: (interval - self.unlockedTimeAllowance))
+        }
+
+        self.defaults.set(date, forKey: Key.applicationDidEnterBackgroundDate.rawValue)
+        self.defaults.synchronize()
+    }
+
+    /// Unlocks the app.
+    ///
+    /// - Important: Should be called if the user session is destroyed, if your app has them, otherwise you risk having the app locked whilst there's no user, preventing people from logging in / signing up.
+    public func unlock() {
+        UserDefaults.standard.removeObject(forKey: Key.applicationDidEnterBackgroundDate.rawValue)
+    }
+
+    /// Requests that the user authenticate with biometrics. Feel free to allow a fallback, like a pincode or password screen.
+
+    public func authenticateWithBiometrics() {
+        var error: NSError?
+        guard self.authenticationContext.canEvaluatePolicy(self.policy, error: &error) else {
+            return
+        }
+
+        self.authenticationContext.evaluatePolicy(self.policy, localizedReason: self.localizedReason) { success, error in
+            DispatchQueue.main.async { [weak self] in
+                guard let strongSelf = self else { return }
+
+                if success {
+                    strongSelf.delegate?.didAuthenticateSuccessfully()
+                    strongSelf.positiveFeedbackGenerator.impactOccurred()
+                    strongSelf.unlock()
+                } else {
+                    strongSelf.negativeFeedbackGenerator.impactOccurred()
+
+                    // This should not actually fail. Not sure why the API doesn't simply return it as an `LAError` directly.
```

Comment:
Apple has a lot of APIs that return stuff like `cancelled` as an error (looking in your direction, `NSURLError`). I also am not a giant fan of this pattern, but I suspect this is at least trying to be internally consistent with that pattern.

Reply:
Yeah, but in this case, AFAIK, it's always an `LAError`, yet it comes as a generic `Error` instead. Must be some ObjC bridging thing.

Reply:
That's definitely likely